### PR TITLE
[RHOAIENG-2977] Artifacts table

### PR DESCRIPTION
**Pull request by jpuzz0**
Closes: [RHOAIENG-2977](https://issues.redhat.com/browse/RHOAIENG-2977)

## Description
Artifacts table, filter toolbar, pagination. 

image
(cc @yannnz)

## Demos
**Filtering**
https://github.com/opendatahub-io/odh-dashboard/assets/96431149/53a04653-b833-49ec-94ec-70d7bcc666a1

**Pagination**
https://github.com/opendatahub-io/odh-dashboard/assets/96431149/a532700a-2edd-4288-a013-9731fc5eef89

## Testing instructions
Navigate to the Artifacts page from the left nav. Using a project with multiple artifacts, verify pagination, and toolbar filtering works as expected.

## How Has This Been Tested?
Cypress does not support intercepting `application/grpc-web+proto` content type requests.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (DaoDaoNoCode) [inline frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts:44]: You can use `response.getArtifactsList()` instead of the conversion by yourself.

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (DaoDaoNoCode) [inline frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts:42]: Is the setting here needed?

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts:44]: They're not the same. That's a different set of data.

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts:42]: It is not needed, I will remove it.

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (DaoDaoNoCode) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Wow, what's the difference because I am using `response.getExecutionList()` to get the executions

Comment (DaoDaoNoCode) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Yes, because I think the intention of the KubeFlow MLMD class is to let you use the method in the class, though it's not documented anywhere. So if you are using `response.getArtifactsList()`, then `artifact.getCreateTimeSinceEpoch()` should give you the same result compared to this line https://github.com/opendatahub-io/odh-dashboard/pull/2744/files#diff-a85ef6d786c1a8509e52597a67c0273d4302a82018ad211b237f1d45bb1f82baR180

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Are you claiming the theoretical intention of the MLMD class and the accessing of values via method call is better in some way than just accessing from the object keys?

Comment (DaoDaoNoCode) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Yes, because this is how they do in the codebase. Also, it matches the type `Artifact` instead of type `Artifact.AsObject`

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: How is it better? They expose AsObject for a reason. Just because they're not using it in their code is not justification for it being better. Unless there's a good reason, I'll be leaving as-is. The consumption of data in the table is more straightforward this way.

File: frontend/src/pages/pipelines/global/experiments/MlmdListContext.tsx
Comment (DaoDaoNoCode) [inline frontend/src/pages/pipelines/global/experiments/MlmdListContext.tsx:1]: It's better to put this file under `/src/concepts/pipelines/context/`, what do you think? It's not an experiment-specific stuff.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/MlmdListContext.tsx:1]: Sure, I can move it.

File: frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx
Comment (DaoDaoNoCode) [inline frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx:51]: Can you make this a hook file like `useDisablePaginationNextButtons(nextPageToken)`? It would be easier to reuse it in the execution table as well.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx:51]: I don't think that's necessary and that's a confusing name, where its also only applicable to MLMD list tables.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx]: I moved the logic inside of `useMlmdListContext` so you get it for free with executions.

**Review by DaoDaoNoCode — APPROVED** [review]
Tested, works well.
/lgtm

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: MLMD generated functions exposes it, not KF. We should use the full functionality of the mlmd resource types which include more than just the data, there are other functions we would not have if we convert the artifact into an object right away. We may not use them know, but maybe in the future we will. I would suggest we keep the main hooks returning the base type, and then further down into implementation, we we can choose to convert that artifact into an object.

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (Gkrumbach07) [inline frontend/src/concepts/pipelines/context/MlmdListContext.tsx:21]: Why do we need a context provider for just the table?

File: frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx
Comment (Gkrumbach07) [inline frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx:93]: can this filter logic be abstracted so the executions can reuse this?

Comment (DaoDaoNoCode) [reply frontend/src/pages/pipelines/global/experiments/artifacts/ArtifactsTable.tsx:93]: I can do this in my PR after this is merged.

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:46]: The pagination arrow has no exposed disabled handler, and due to the way data is only fetched per page for MLMD resources, this was necessary to disable the next arrow on the last page.

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:21]: I added the context so that Juntao can use it for Executions as well. It would be for any list table using MLMD resources.

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Why would we want anything other than the data here? Is that not what every other API fetching hook we have does?

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:46]: can you do something where once you have a next page token, you set the item count to some defiend value. then it will disable on its own 

for ref this is how the page size is calculated
```tsx
 // when itemCount is not known let's set lastPage as page+1 as we don't know the total count
itemCount || itemCount === 0 ? Math.ceil(itemCount / perPage) || 0 : page + 1;
```

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: I am trying to keep consistency across the feature. If we are using one type everywhere else, but then another version of the type in one other spot, it can get confusing. And you can still use that object further down in implementation.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: Where would we use Artifact as a non-object form and why?

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: I would like to know what the answers to these questions are, but since this seems to be blocking the PR, I went ahead and updated my hook to return the list in a non-object form, and am just having to iterate over all artifacts now for the table, which is more code than what was there prior.

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:46]: OK in that case, can you log a PF issue and add that too the code as a comment please. This really should not be a way we do it, but if there is no other option then its fine but should be tracked

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:21]: closing as convo is in slack

File: frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts
Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/experiments/artifacts/useGetArtifactsList.ts]: We are not using it now, but for consistency across all of our other MLMD hooks

File: frontend/src/concepts/pipelines/context/MlmdListContext.tsx
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/context/MlmdListContext.tsx:47]: I created the issue and added a comment with a link to that PF issue.

**Comment by Gkrumbach07** [comment]
all good to me.

/approve

**Review by DaoDaoNoCode — COMMENTED** [review]
/lgtm

**Pull request merged**
